python/ray/data/random_access_dataset.py

```diff
@@ -0,0 +1,262 @@
+import bisect
+import logging
+import random
+import time
+from collections import defaultdict
+from typing import List, Any, Optional, TYPE_CHECKING
+
+import ray
+from ray.types import ObjectRef
+from ray.data.block import T, BlockAccessor
+from ray.data.impl.remote_fn import cached_remote_fn
+
+if TYPE_CHECKING:
+    from ray.data import Dataset
+
+logger = logging.getLogger(__name__)
+
+
+class RandomAccessDataset(object):
+    """A class that provides distributed, random access to a Dataset.
+
+    See: ``Dataset.to_random_access_dataset()``.
+    """
+
+    def __init__(
+        self,
+        dataset: "Dataset[T]",
+        key: str,
+        num_workers: int,
+    ):
+        """Construct a RandomAccessDataset (internal API).
+
+        The constructor is not part of the Datasets API. Use
+        ``dataset.to_random_access_dataset()`` to construct a RandomAccessDataset.
+        """
+        self._format = dataset._dataset_format()
+        if self._format not in ["arrow", "pandas"]:
+            raise ValueError("RandomAccessDataset only supports Arrow-format datasets.")
+
+        start = time.perf_counter()
+        logger.info("[setup] Indexing dataset by sort key.")
+        sorted_ds = dataset.sort(key)
+        get_bounds = cached_remote_fn(_get_bounds)
+        blocks = sorted_ds.get_internal_block_refs()
+
+        logger.info("[setup] Computing block range bounds.")
+        bounds = ray.get([get_bounds.remote(b, key, self._format) for b in blocks])
+        self._valid_blocks = []
+        self._lower_bound = None
+        self._upper_bounds = []
+        for i, b in enumerate(bounds):
+            if b:
+                self._valid_blocks.append(blocks[i])
+                if self._lower_bound is None:
+                    self._lower_bound = b[0]
+                self._upper_bounds.append(b[1])
+
+        logger.info("[setup] Creating {} random access workers.".format(num_workers))
+        self._workers = [
+            _RandomAccessWorker.options(scheduling_strategy="SPREAD").remote(
+                key, self._format
+            )
+            for _ in range(num_workers)
+        ]
+        (
+            self._block_to_workers_map,
+            self._worker_to_blocks_map,
+        ) = self._compute_block_to_worker_assignments()
+
+        logger.info(
+            "[setup] Worker to blocks assignment: {}".format(self._worker_to_blocks_map)
```

Comment:
What's the story around node failure: e.g. an actor worker node fails, a data block node fails.

Reply:
We don't support fault tolerance. This could be something we re-visit in the future with restartable actors, but I don't think it's high priority.

Reply:
Should we document this?

Reply:
Added to docs.